Add a unit test for idempotent applys to the TPR entries.

ahakanbaba opened the pull request:
The test in apply_test follows the general pattern of other tests.
We load from a file in test/fixtures and mock the API server in the
function closure in the HttpClient call.
The apply operation expects a last-modified-configuration annotation.
That is written verbatim in the test/fixture file.

References #40841

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Adds one unit test for TPR's using applies. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
References: 
https://github.com/kubernetes/features/issues/95
https://github.com/kubernetes/kubernetes/issues/40841#issue-204769102


**Special notes for your reviewer**:

I am not super proud of the tpr-entry name. 
But I feel like we need to call the two objects differently. 
The one which has Kind:ThirdPartyResource 
and the one has Kind:Foo. 

Is the name "ThirdPartyResource" used interchangeably for both ? I used tpr-entry for the Kind:Foo object.

Also I !assume! this is testing an idempotent apply because the last-applied-configuration annotation is the same as the object itself. 

This is the state I see in the logs of kubectl if I do a proper idempotent apply of a third party resource entry. 

I guess I will know more once I start playing around with apply command that change TPR objects. 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

ahakanbaba committed:
Add a unit test for applies and idempotent applys to the TPR entries.
The tests in apply_test follows the general pattern of other tests.
We load from a file in test/fixtures and mock the API server in the
function closure in the HttpClient call.
In PATCH request rount-tripper we check that the kubectl apply
implementation worked as expected.

References #40841

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41151)
<!-- Reviewable:end -->

ahakanbaba commented:
@ymqytw @liggitt 
This is my first commit regarding the test plan in https://github.com/kubernetes/kubernetes/issues/40841#issue-204769102

Please take a look at the change. Depending on your comments I can add the other tests in a similar way.

liggitt commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -701,5 +701,21 @@ func testDynamicResources() []*discovery.APIGroupResources {
+				},
+				PreferredVersion: metav1.GroupVersionForDiscovery{
+					GroupVersion: "unit-test.test.com/v1",
+					Version:      "v1beta1"},
```
v1?

ahakanbaba replied:
Yep :) Good catch thanks. 
The test was still passing like this though.

liggitt commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -701,5 +701,21 @@ func testDynamicResources() []*discovery.APIGroupResources {
+			},
+			VersionedResources: map[string][]metav1.APIResource{
+				"v1": {
+					{Name: "testtprs", Namespaced: true, Kind: "TestTpr"},
```
this is not TPR-specific, prefer `{Name: "widgets", Namespaced: true, Kind: "Widget"}` or something

liggitt commented on test/fixtures/pkg/kubectl/cmd/apply/tpr-entry.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: "unit-test.test.com/v1"
+kind: TestTpr
```
again, not TPR-specific... this is testing the ability to apply generic objects not compiled into kubectl

ahakanbaba replied:
OK I can make it widget. no problem.

ahakanbaba commented:
I would be also interested in the real value of this test.
Maybe I should cherrypick this change on master's state before #40666 merged and see that this test really fails ?

liggitt commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -531,3 +534,89 @@ func TestApplyNULLPreservation(t *testing.T) {
+	errBuf := bytes.NewBuffer([]byte{})
+
+	cmd := NewCmdApply(f, buf, errBuf)
+	cmd.Flags().Set("filename", filenameIdempotentTPREntry)
```
you should not give the apply command a fixture that already has a last-applied annotation in it. that should only exist server-side.

there are two testcases here:

1. create case:
  * apply -f file-without-annotation.yaml
  * server responds with 404
  * create API call is made
  * verify submitted content matches the fixture content *with an additional last-applied annotation added*

2. idempotent apply:
  * apply -f file-without-annotation.yaml
  * server responds with 200 response with object matching the fixture *with a last-applied annotation added*
  * apply no-ops (I forget if it sends an empty patch or skips patching in this case)
  * verify patch content is empty and content-type is generic json merge patch

ahakanbaba replied:
Understood. 

I ran this scenario with a compiled kubectl against a running api server. 
The request of the PATCH differs in the unit test and the real example. 

In Unit Test: 
```
I0208 21:53:58.602190   21586 request.go:991] Request Body: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"}}}
I0208 21:53:58.602218   21586 request.go:991] Response Body: {"apiVersion":"unit-test.test.com/v1","key":"value","kind":"Widget","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"name\":\"widget\",\"namespace\":\"test\"}}\n"},"name":"widget","namespace":"test"}}
PASS
```

In real version: 
```
I0208 21:40:17.639077    5803 request.go:991] Request Body: {"metadata":{"annotations":{}}}
I0208 21:40:17.639127    5803 round_trippers.go:398] curl -k -v -XPATCH  -H "Accept: application/json" -H "Content-Type: application/merge-patch+json" -H "User-Agent: kubectl/v0.0.0 (linux/amd64) kubernetes/$Format" https://localhost:
6443/apis/unit-test.test.com/v1/namespaces/test/widgets/widget
I0208 21:40:17.641624    5803 round_trippers.go:417] PATCH https://localhost:6443/apis/unit-test.test.com/v1/namespaces/test/widgets/widget 200 OK in 2 milliseconds
I0208 21:40:17.641641    5803 round_trippers.go:423] Response Headers:
I0208 21:40:17.641645    5803 round_trippers.go:426]     Content-Type: application/json
I0208 21:40:17.641647    5803 round_trippers.go:426]     Content-Length: 530
I0208 21:40:17.641649    5803 round_trippers.go:426]     Date: Wed, 08 Feb 2017 21:40:17 GMT
I0208 21:40:17.641667    5803 request.go:991] Response Body: {"apiVersion":"unit-test.test.com/v1","key":"value","kind":"Widget","metadata":{"name":"widget","namespace":"test","selfLink":"/apis/unit-test.test.com/v1/namespaces/test/wi
dgets/widget","uid":"2c81844a-ee47-11e6-a0f6-0800275ae9e7","resourceVersion":"117","creationTimestamp":"2017-02-08T21:40:13Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\"
,\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"}}}
dget "widget" configured
```

The unit test adds the kubectl.kubernetes.io/last-applied-configuration to the annotations to the PATCH request.
Maybe you have any idea what may be the reason ?

ahakanbaba commented:
@liggitt I think I have addressed all your comments. 
Can you please take another look ?

liggitt commented on docs/api-reference/certificates/v1alpha1/definitions.html:
```diff
@@ -0,0 +1,1318 @@
+<!DOCTYPE html>
```
looks like a spurious diff

ahakanbaba replied:
Sorry about that. :( 
Deleted the file. Still do not understand how it got created though.

ahakanbaba commented:
@liggitt  Hopefully for one more final time, take a look at it ?

liggitt commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -113,18 +125,17 @@ func readServiceFromFile(t *testing.T, filename string) *api.Service {
-	original, err := runtime.Encode(testapi.Default.Codec(), originalObj)
-	if err != nil {
-		t.Fatal(err)
+	if originalLabels != nil {
```
if this is nil, I'd expect the labels to be initialized:
```
if originalLabels == nil {
  originalLabels = make(map[string]string)
}
```

ahakanbaba replied:
@liggitt 
Are you sure adding the DELETE_ME label no matter what is the right thing? 
1) The test fails if I make the modification you have suggested.  (The PATCH request differs from the empty request) 
```
[vagrant@k8s-env cmd]$ go test -v -run TestApplyIdempotentTPREntry
=== RUN   TestApplyIdempotentTPREntry
--- FAIL: TestApplyIdempotentTPREntry (0.00s)
	apply_test.go:614: Unexpected Patch. Found unexpected annotation: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"},"labels":null}}
FAIL
```

2) The GET request's response will differ from the actual kubectl execution. 
I execute the following command on a running api server 
` kubectl.sh --kubeconfig=/var/run/kubernetes/admin.kubeconfig  apply -f widget.yaml -v=10`
And the logs show the following info about the GET request 
```
I0211 02:57:19.899574   22691 round_trippers.go:417] GET https://localhost:6443/apis/unit-test.test.com/v1/namespaces/test/widgets/widget 200 OK in 1 milliseconds
I0211 02:57:19.899585   22691 round_trippers.go:423] Response Headers:
I0211 02:57:19.899590   22691 round_trippers.go:426]     Date: Sat, 11 Feb 2017 02:57:19 GMT
I0211 02:57:19.899593   22691 round_trippers.go:426]     Content-Type: application/json
I0211 02:57:19.899596   22691 round_trippers.go:426]     Content-Length: 532
I0211 02:57:19.899614   22691 request.go:991] Response Body: {"apiVersion":"unit-test.test.com/v1","key":"value","kind":"Widget","metadata":{"name":"widget","namespace":"test","selfLink":"/apis/unit-test.test.com/v1/namespaces/test/widgets/widget","uid":"3e03a56e-f005-11e6-8716-0800275ae9e7","resourceVersion":"14220","creationTimestamp":"2017-02-11T02:53:18Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"}}}
```

As you can see the response of the GET request does not have the DELETE_ME label. Is that wise a different behavior to the unit test ?

liggitt replied:
ok, I dug in and see what is happening. 

first, `annotateRuntimeObject` expects an object that has labels set in its metadata (which is why you were getting a nil panic), and is intended to be paired with `validatePatchApplication`. it adds the "DELETE_ME" label to make sure that reapplying the original object (which is supposed to have some labels set) removes the added label. That's a useful test to run, but it is not the same test as your idempotent apply test. I would not modify `annotateRuntimeObject` to fit your idempotent testcase.

For your idempotent testcase, I would suggest two stub files, one representing clientside data and one serverside data (complete with last-applied annotation). You can just feed the client-side stub to the apply command, and the server-side stub to the server, and verify an empty patch is sent.

With that approach, no changes are needed to the existing apply tests, which is nice. Opened https://github.com/ahakanbaba/kubernetes/pull/1 with those suggestions if you are amenable.

liggitt commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -136,9 +147,16 @@ func annotateRuntimeObject(t *testing.T, originalObj, currentObj runtime.Object,
 	}
-	currentAnnotations[annotations.LastAppliedConfigAnnotation] = string(original)
+
+	originalAccessor.SetAnnotations(map[string]string{})
```
don't do this unconditionally, this could be stomping data we care about

ahakanbaba replied:
ok I have added a conditional. 

Without this empty  annotation, the PATCH request does not become empty. 
In kubectl execution the PATH request is actually empty 
`Request Body: {"metadata":{"annotations":{}}}`
However in this unit test, the request differs. 
`{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"}}}` 

I made a previous comment on this PR  about that and asked you whether you new the reason of it. 

In this  [commit](https://github.com/kubernetes/kubernetes/pull/41151/commits/ed1da06d7692f9c5a75f36f36edc6ee910110fbd) submit message I also show that bunch of other PATCH requests going to the API server also contain the empty annotation.

liggitt commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -136,9 +147,16 @@ func annotateRuntimeObject(t *testing.T, originalObj, currentObj runtime.Object,
+		t.Fatal(err)
+	}
+
+	currentAnnotations[annotations.LastAppliedConfigAnnotation] = string(originalWithEmptyAnno)
```
can you describe why this change is needed?

ahakanbaba replied:
Do you want me to add a comment ? 
I do not know the actual reason in the implementation. 
I have looked into the logs of kubectl and found this difference. Adding the same field here made the PATCH requests the same. 
Does writing just that in a comment add enough information to the code? 

Do you have an intelligent blob for me to add as a comment ?

liggitt commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -113,18 +125,17 @@ func readServiceFromFile(t *testing.T, filename string) *api.Service {
 }
 
-func annotateRuntimeObject(t *testing.T, originalObj, currentObj runtime.Object, kind string) (string, []byte) {
+func annotateRuntimeObject(t *testing.T, originalObj, currentObj runtime.Object,
```
no newline

liggitt commented:
go ahead and squash, then LGTM

ahakanbaba commented:
ok Thanks. 
I see now what was the story about the DELETE_ME tag. 
The tests essentially check that the apply command results in deletion of unrelated tags. 
Do you mind if I add a comment in the annotateRuntimeObject before DELETE_ME insertion ?

liggitt commented:
> Do you mind if I add a comment in the annotateRuntimeObject before DELETE_ME insertion ?

Not at all... describing what it is for, the requirements for an incoming object (have some labels), and that it is designed to be paired with validatePatchApplication would be really helpful to the next people

ahakanbaba commented:
Ok Added the comment. Feel free to review the wording.

ahakanbaba commented:
@liggitt Also squashed. 
Thanks for all the back and forth and the useful comments.

liggitt commented:
/lgtm /approve

ahakanbaba commented:
>  they should reply with @k8s-bot ok to test on its own line.  
DO you need to reply with `@k8s-bot ok to test ` ?

liggitt commented:
@k8s-bot ok to test

ahakanbaba commented:
@liggitt I am seeing build errors 

```I0211 06:45:48.603] Makefile:119: recipe for target 'verify' failed
W0211 06:45:48.703] make: *** [verify] Error 1
W0211 06:45:51.297] Traceback (most recent call last):
W0211 06:45:51.297]   File "/var/lib/jenkins/workspace/pull-kubernetes-verify/./test-infra/jenkins/../scenarios/kubernetes_verify.py", line 107, in <module>
W0211 06:45:51.298]     main(ARGS.branch, ARGS.script, ARGS.force)
W0211 06:45:51.298]   File "/var/lib/jenkins/workspace/pull-kubernetes-verify/./test-infra/jenkins/../scenarios/kubernetes_verify.py", line 91, in main
W0211 06:45:51.298]     'bash', '-c', 'cd kubernetes && %s' % script,
W0211 06:45:51.298]   File "/var/lib/jenkins/workspace/pull-kubernetes-verify/./test-infra/jenkins/../scenarios/kubernetes_verify.py", line 51, in check
W0211 06:45:51.298]     subprocess.check_call(cmd)
W0211 06:45:51.299]   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
W0211 06:45:51.299]     raise CalledProcessError(retcode, cmd)
W0211 06:45:51.299] subprocess.CalledProcessError: Command '('docker', 'run', '--rm=true', '--privileged=true', '-v', '/var/run/docker.sock:/var/run/docker.sock', '-v', '/etc/localtime:/etc/localtime:ro', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-verify/go/src/k8s.io/kubernetes:/go/src/k8s.io/kubernetes', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-verify/go/src/k8s.io/kubernetes/_artifacts:/workspace/artifacts', '-e', 'KUBE_FORCE_VERIFY_CHECKS=n', '-e', 'KUBE_VERIFY_GIT_BRANCH=master', '-e', 'REPO_DIR=/var/lib/jenkins/workspace/pull-kubernetes-verify/go/src/k8s.io/kubernetes', 'gcr.io/k8s-testimages/kubekins-test:1.6-v20161205-ad918bc', 'bash', '-c', 'cd kubernetes && ./hack/jenkins/verify-dockerized.sh')' returned non-zero exit status 2
E0211 06:45:51.304] Build failed
I0211 06:45:51.305] process 12312 exited with code 1 after 27.2m
E0211 06:45:51.305] FAIL: pull-kubernetes-verify
```

Not sure how to debug these. 
What is the distribution of these Jenkins machines. 

Do the PR owners have access to them ?

liggitt commented:
```
... skipping 1931 lines ...
I0211 06:23:50.729] >         "//vendor:k8s.io/apimachinery/pkg/apis/meta/v1/unstructured",
W0211 06:23:50.829] DRY-RUN: wrote "pkg/kubectl/cmd/BUILD"
W0211 06:23:51.047] 
W0211 06:23:51.047] 1 BUILD files not up-to-date.
I0211 06:23:51.148] 
I0211 06:23:51.148] Run ./hack/update-bazel.sh
I0211 06:23:51.148] FAILED   hack/make-rules/../../hack/verify-bazel.sh	5s
```

ahakanbaba commented:
Thanks for pointing it out. It looks like it was obviously at the top. 

I do not know much about bazel, gazel usage.
Running ./hack/update-bazel.sh 

causes bunch of errors in my development environment
```
......
extract err: cannot find package "k8s.io/apiserver/pkg/storage/storagebackend" in any of:
	/usr/local/go/src/k8s.io/apiserver/pkg/storage/storagebackend (from $GOROOT)
	/home/vagrant/gopath/src/k8s.io/apiserver/pkg/storage/storagebackend (from $GOPATH)
extract err: cannot find package "k8s.io/apiserver/pkg/endpoints/request" in any of:
	/usr/local/go/src/k8s.io/apiserver/pkg/endpoints/request (from $GOROOT)
	/home/vagrant/gopath/src/k8s.io/apiserver/pkg/endpoints/request (from $GOPATH)
extract err: cannot find package "k8s.io/apiserver/pkg/registry/generic" in any of:
	/usr/local/go/src/k8s.io/apiserver/pkg/registry/generic (from $GOROOT)
	/home/vagrant/gopath/src/k8s.io/apiserver/pkg/registry/generic (from $GOPATH)
extract err: cannot find package "k8s.io/apiserver/pkg/registry/generic/registry" in any of:
	/usr/local/go/src/k8s.io/apiserver/pkg/registry/generic/registry (from $GOROOT)
	/home/vagrant/gopath/src/k8s.io/apiserver/pkg/registry/generic/registry (from $GOPATH)
extract err: cannot find package "k8s.io/apiserver/pkg/storage" in any of:
.....
````

The errors may be caused because of something missing in my dev environment. Where do you develop and build kubernetes code ? So far I have been using vagrant vm in a mac using this https://github.com/ghodss/kubernetes-macosx-development. Maybe I need to move to a more popular dev environment. 

Although there are bunch of errors, the script exits with success. Leaves bunch of files as modified.

```
......
	modified:   cmd/cloud-controller-manager/app/BUILD
	modified:   cmd/cloud-controller-manager/app/options/BUILD
	modified:   cmd/hyperkube/BUILD
	modified:   cmd/kube-aggregator/pkg/apiserver/BUILD
	modified:   cmd/kube-aggregator/pkg/cmd/server/BUILD
	modified:   cmd/kube-aggregator/pkg/registry/apiservice/BUILD
	modified:   cmd/kube-aggregator/pkg/registry/apiservice/etcd/BUILD
	modified:   cmd/kube-apiserver/BUILD
	modified:   cmd/kube-apiserver/app/BUILD
	modified:   cmd/kube-apiserver/app/options/BUILD
.....
Many more 
```

I am not sure all this is right. 
We have added two functions to one unit test file. Modified couple of lines in fake.go and added two test files.

Are all these modifications expected ?

liggitt commented:
only a single modification is expected:

```
diff --git a/pkg/kubectl/cmd/BUILD b/pkg/kubectl/cmd/BUILD
index edb6fd6..5456212 100644
--- a/pkg/kubectl/cmd/BUILD
+++ b/pkg/kubectl/cmd/BUILD
@@ -199,6 +199,7 @@ go_test(
         "//vendor:k8s.io/apimachinery/pkg/api/meta",
         "//vendor:k8s.io/apimachinery/pkg/api/resource",
         "//vendor:k8s.io/apimachinery/pkg/apis/meta/v1",
+        "//vendor:k8s.io/apimachinery/pkg/apis/meta/v1/unstructured",
         "//vendor:k8s.io/apimachinery/pkg/runtime",
         "//vendor:k8s.io/apimachinery/pkg/runtime/schema",
         "//vendor:k8s.io/apimachinery/pkg/runtime/serializer/json",
```

ymqytw commented:
@ahakanbaba Please run `hack/update-bazel.sh` to update bazel.

ymqytw commented:
@k8s-bot cross build this
@k8s-bot kubemark e2e test this

ahakanbaba commented:
@ymqytw sorry for my embarrassingly long delay on this. 
I was pulled away to some other tasks.

liggitt commented on vendor/BUILD:
```diff
@@ -17120,3 +17110,13 @@ go_library(
 )
+
+go_library(
+    name = "github.com/ClusterHQ/flocker-go",
```
This looks unrelated.

ahakanbaba replied:
I ran update-bazel.sh from my mac and these were the changes. 
Do you want me to delete these ?

ahakanbaba replied:
I see what is going on. 
The location of this code section has changed. 

There is a deletion in the same file for the same code. 


This is what I get. 
```
[vagrant@k8s-env kubernetes]$ git status
# On branch tpr-unit-tests
nothing to commit, working directory clean
[vagrant@k8s-env kubernetes]$ ./hack/update-bazel.sh
[vagrant@k8s-env kubernetes]$
[vagrant@k8s-env kubernetes]$
[vagrant@k8s-env kubernetes]$ git status
# On branch tpr-unit-tests
nothing to commit, working directory clean
```

update-bazel does not cause any more changes to me.

ahakanbaba commented:
In my development environment the verify-bazel works. 

[vagrant@k8s-env kubernetes]$ ./hack/verify-bazel.sh
[vagrant@k8s-env kubernetes]$ echo $?
0

I am afraid I need some help with these test failures

ahakanbaba commented:
Maybe the bazel version is different ? 
I have used the latest 0.4.4

ahakanbaba commented:
@liggitt All tests pass .
Can you LGTM it ?

liggitt commented:
/lgtm

ahakanbaba commented:
@liggitt 
I do not know what the last remaining check is. 

code-review/reviewable — 5 files, 9 discussions left (ahakanbaba, liggitt)

How can we address the last check ?

liggitt commented:
you can ignore it, it's reviewable stuck on a review. if the submit queue has a green check, it's happy and you're just in the (long, long) queue

ahakanbaba commented:
@liggitt I had to rebase and merge a spacing merge conflict. 
Unfortunately you have to LGTM again. 

Did I lose my previous spot on the SubmitQueue because of this rebase ? 

Is there a way to see my spot on the mergeQ ?

liggitt commented:
/lgtm

liggitt commented:
once you're in the queue, the PR shows up in http://submit-queue.k8s.io
